### PR TITLE
Editorial: Restrict IsValidTimeZoneName case folding to ASCII

### LINKED ISSUE #1919: IsValidTimeZoneName "UTC" checking should limit case folding to ASCII
[IsValidTimeZoneName](https://tc39.es/proposal-temporal/#sec-isvalidtimezonename) currently uppercases the input "according to the Unicode Default Case Conversion algorithm", but it would be inappropriate for a hypothetical non-ASCII string that case-folds to "UTC" to be recognized as the UTC time zone.

### PR DESCRIPTION
**Pull request by gibson042**
Fixes #1919

### CODE REVIEW

File: spec/intl.html
Comment (gibson042) [inline spec/intl.html:28]: This corrects a current mistake in ECMA-402, which overrides "Etc/UTC" and "Etc/GMT" (which are canonical names for distinct time zones) with "UTC" (which is an alias of the former).

File: spec/intl.html
Comment (gibson042) [inline spec/intl.html:32]: It is implicit in the algorithms of ECMA-402 IsValidTimeZoneName and CanonicalizeTimeZoneName that only IANA time zone names are valid. This makes it explicit.

Whether this is appropriate, and whether a _non_-402 implementation should be allowed to support a non-IANA time zone name (as in the current state of Temporal) is not answered here.

File: spec/intl.html
Comment (gibson042) [inline spec/intl.html:35]: Looking up built-in time zones in ECMA-262 IsValidTimeZoneName removes the need to define that algorithm in ECMA-402.

File: spec/timezone.html
Comment (gibson042) [inline spec/timezone.html:71]: It would also be possible to remove the ECMA-402 override of CanonicalizeTimeZoneName by including the tzdb steps and the "UTC" overrides here in ECMA-262, but I don't know if that is considered appropriate.

**Review by ptomato — APPROVED** [review]
This looks like a good change. Although I'm not so confident about my mental model of how things should be divided between ECMA-262 and 402.

@ryzokuken @sffc It would be good to get your review on this as well, for that reason. @FrankYFTang also your perspective as implementor and ECMA-402 contributor.

**Review by sffc — COMMENTED** [review]
Seems fine.  Procedurally, I think it would be slightly cleaner to make this a standalone normative PR against ECMA-402.

**Comment by gibson042** [comment]
> I think it would be slightly cleaner to make this a standalone normative PR against ECMA-402.

@sffc Done at https://github.com/tc39/ecma402/pull/624, which ended up being editorial. This PR can be simplified after that lands, but will look different based on whether or not a non-ECMA-402 ECMA-262 Temporal implementation is allowed to support a built-in time zone that is absent from the IANA database (which is not allowed of an ECMA-402 implementation).

**Comment by ptomato** [comment]
https://github.com/tc39/ecma402/pull/624 is merged now, so I think this can be edited somewhat.

File: spec/timezone.html
Comment (gibson042) [inline spec/timezone.html:78]: There's admittedly a lot of duplication here. What would you think of removing IsValidTimeZoneName from both ECMA-262 and ECMA-402 by merging it into CanonicalizeTimeZoneName (which every invocation is already followed by) such that the latter would return undefined when given an invalid time zone name rather than requiring that the name has already been validated (and if there's approval, would that be in scope or out of scope for this PR)?

File: spec/timezone.html
Comment (gibson042) [inline spec/timezone.html:85]: I'd love to remove _all_ the ECMA-402-specific operation overrides by slightly rewording the already-existing constraints that limit built-in time zone names to those published by IANA and incorporating the Link following into ECMA-262 CanonicalizeTimeZoneName (presumably Normative Optional), but I'm not quite sure how to pull it off. Maybe it will be more clear if IsValidTimeZoneName is removed from both as I suggested above.

**Review by gibson042 — COMMENTED** [review]
Done.

File: spec/timezone.html
Comment (Ms2ger) [inline spec/timezone.html:34]: Hmm, I thought we were assuming non-402 implementations would only support UTC and maybe offsets? I might be misremembering, though.

File: spec/timezone.html
Comment (Ms2ger) [inline spec/timezone.html:33]: Is that because the offset time zones are handled separately? It might be worth calling that out.

File: spec/timezone.html
Comment (Ms2ger) [inline spec/timezone.html:85]: Do we want to use "should" in a note?

**Comment by FrankYFTang** [comment]
I have hard time w/ https://github.com/tc39/test262/pull/3346 . I am not sure what is the right fix in the spec needed for those tests to pass but right now what I see is the name is not taking TimeZoneUTCOffsetName and the grammar could be changed by https://github.com/tc39/proposal-temporal/pull/1941 but that is not enough because step 7-a and 7-b of ParseTemporalTimeZoneString will throw. I have not carefully read all detail of this PR and I am not sure this PR will solve all the remaining issues but surely either the calling place in 7a or inside IsValidTimeZoneName need to be changed to solve that.

File: spec/timezone.html
Comment (gibson042) [reply spec/timezone.html:33]: It's just inherent in the behavior specified at https://tc39.es/ecma402/#sec-time-zone-names that any purported time zone name not present in the IANA data is invalid. So e.g. Intl.DateTimeFormat accepts "Etc/GMT+5" as `timeZone` but rejects "-05:00" with a RangeError.

File: spec/timezone.html
Comment (gibson042) [reply spec/timezone.html:85]: I'm definitely open to other ideas. Are you looking for a rephrase like "implementations are encouraged to…"?

File: spec/timezone.html
Comment (ptomato) [reply spec/timezone.html:34]: I think the idea was more like, non-402 implementations support at least UTC and offsets (which don't go through this path, though) and are free to support any additional subset of IANA zones or any other wacky zones that they please, although they are encouraged to do the same as 402. We affirmed this in yesterday's Temporal champions meeting.

File: spec/timezone.html
Comment (ptomato) [inline spec/timezone.html:55]: Are you sure this can be editorial? It seems to me that this would be normative if there were, as you described in #1919, "a hypothetical non-ASCII string that case-folds to `"UTC"`", but maybe no such string exists?

**Review by ptomato — COMMENTED** [review]
Is it the case that anything needs to change in this PR according to the rough consensus we reached in yesterday's champions meeting? For example, do we now need to decouple IsValidTimeZoneName from 402's IsValidTimeZoneName?

File: spec/timezone.html
Comment (gibson042) [reply spec/timezone.html:55]: Correct, that's exactly why I believe this to be editorial.

**Comment by gibson042** [comment]
> Is it the case that anything needs to change in this PR according to the rough consensus we reached in yesterday's champions meeting? For example, do we now need to decouple IsValidTimeZoneName from 402's IsValidTimeZoneName?

Yes, I think so.

**Comment by ptomato** [comment]
> > Is it the case that anything needs to change in this PR according to the rough consensus we reached in yesterday's champions meeting? For example, do we now need to decouple IsValidTimeZoneName from 402's IsValidTimeZoneName?
> 
> Yes, I think so.

On second thought, I think it's probably better to put that into its own issue (#1996) so that we don't keep this PR up in the air longer than necessary. I can rebase this today.

**Comment by ptomato** [comment]
I guess it would be best in that case to remove "Editorial: Add implementation-defined built-in time zone lookup steps" since that takes things more in the direction of unifying 402 and 262 IsValidTimeZone, rather than splitting them. It's a bit difficult to separate but I'll continue working on this tomorrow.

File: spec/timezone.html
Comment (ptomato) [reply spec/timezone.html]: I think we should consider that out of scope for this PR, but it seems like a good idea.

File: spec/timezone.html
Comment (ptomato) [reply spec/timezone.html]: If we are going in the direction of decoupling them, then this seems no longer applicable.

File: spec/timezone.html
Comment (ptomato) [reply spec/timezone.html]: We agreed at the champions meeting that we would like ECMA-402 to support offset time zones, so I removed this for now.

File: spec/timezone.html
Comment (ptomato) [reply spec/timezone.html]: I removed this for now, since we are going to decouple the 262 and 402 operations.

**Comment by ptomato** [comment]
@gibson042 I've kept the changes that I removed from this PR in a separate branch, https://github.com/tc39/proposal-temporal/tree/gibson042/implementation-defined in case you want to reuse some parts later.

**Pull request merged**
